src/lib/update-clients.ts

```diff
@@ -0,0 +1,69 @@
+import axios from "axios";
+import { GET_OAUTH_CLIENT, PUT_OAUTH_CLIENT } from "./constants";
+
+export interface OAuthClientConfig {
+  urls: string[];
+  name: string;
+}
+interface ClientParams {
+  redirectUris: OAuthClientConfig[];
+  originsToAdd: string[];
+  AM_URL: string;
+  ssoToken: string;
+  cookieName: string;
+  realm: string;
+}
+async function updateRedirectUris({
+  AM_URL,
+  cookieName,
+  originsToAdd,
+  realm,
+  redirectUris,
+  ssoToken,
+}: ClientParams): Promise<void> {
+  const request = axios.create({
+    baseURL: AM_URL,
+    headers: {
+      "Content-Type": "application/json",
+      "accept-api-version": "protocol=2.1, resource=1.0",
+      [cookieName]: ssoToken,
+    },
+  });
+
+  for (const { urls, name } of redirectUris) {
+    const amConfigUrl = GET_OAUTH_CLIENT.url(realm, name);
+
+    try {
+      const { data: config } = await request.get(amConfigUrl);
+
+      const redirectionUris = originsToAdd.reduce(
+        (acc: Array<string>, curr) =>
+          acc.concat(
+            // sanitize urls for extra /'s'
+            urls.map((url) => `${curr}${url}`.replace(/([^:]\/)\/+/g, "$1"))
+          ),
+        []
+      );
+
+      const set = new Set(config.coreOAuth2ClientConfig.redirectionUris.value);
+
+      for (const val of redirectionUris) {
+        set.add(val);
+      }
+
+      const finalRedirectionUris = Array.from(set);
```

Comment:
There's quite a lot of data transformation happening here, and I'm not sure how to grok it. Maybe some code comments that have sample data at each step of this flow, so I can visualize the code's intention ... its inputs and outpots?